### PR TITLE
add more receive functions

### PR DESCRIPTION
**Pull request by chessai**

### CODE REVIEW

File: src/Socket/Stream/IPv4.hs [hunk 1]
```diff
@@ -1280,6 +1283,47 @@ receiveBetweenMutableByteArraySlice !conn@(Connection fd) !minBytes !maxBytes !m
...
+     --   @'Left' 'ReceiveInterrupted'@.
+  -> Connection -- ^ Connection
+  -> Int -- ^ Maximum number of bytes to receive
+  -> Int -- ^ Offset into the buffer
+  -> IO (Either (ReceiveException 'Interruptible) ByteString) -- ^ Either a socket exception or the number of ByteString we allocated
+interruptibleReceiveBoundedByteStringSlice !abandon !conn !total !off = do
+  marr@(MutableByteArray marr#) <- PM.newPinnedByteArray total
+  interruptibleReceiveBoundedMutableByteArraySlice abandon conn total marr off >>= \case
+    Left err -> pure (Left err)
+    Right _ -> pure (Right (BI.PS (FP.ForeignPtr (byteArrayContents# (unsafeCoerce# marr#)) (FP.PlainPtr marr#)) 0 total))
```
Comment (andrewthad) [inline src/Socket/Stream/IPv4.hs]: The length in the bytestring is wrong. The value you actually want for this is the value in the `Right` data constructor.

File: src/Socket/Stream/IPv4.hs [hunk 2]
```diff
@@ -1280,6 +1283,47 @@ receiveBetweenMutableByteArraySlice !conn@(Connection fd) !minBytes !maxBytes !m
...
+  -> Connection -- ^ Connection
+  -> MutableByteArray RealWorld -- ^ Buffer in which the data is going to be stored
+  -> IO (Either (ReceiveException 'Interruptible) Int) -- ^ Either a socket exception or the number of bytes read
+interruptibleReceiveMutableByteArray !abandon !conn !marr
+  = interruptibleReceiveMutableByteArraySlice abandon conn marr 0
+
+-- | Receive bytes using the given array and
+--   starting at the given offset. This can be interrupted by the
+--   completion of an 'STM' transaction.
+interruptibleReceiveMutableByteArraySlice ::
```
Comment (andrewthad) [inline src/Socket/Stream/IPv4.hs]: This function is missing an argument that describes the length of the slice.

File: src/Socket/Stream/IPv4.hs [hunk 3]
```diff
@@ -1280,6 +1283,47 @@ receiveBetweenMutableByteArraySlice !conn@(Connection fd) !minBytes !maxBytes !m
...
+     TVar Bool
+     -- ^ Interrupted. If this becomes 'True' give up and return
+     --   @'Left' 'ReceiveInterrupted'@.
+  -> Connection -- ^ Connection
+  -> MutableByteArray RealWorld -- ^ Buffer in which the data is going to be stored
+  -> Int -- ^ Offset into the buffer
+  -> IO (Either (ReceiveException 'Interruptible) Int) -- ^ Either a socket exception or the number of bytes read
+interruptibleReceiveMutableByteArraySlice !abandon !conn !marr !off = do
+  let !sz = PM.sizeofMutableByteArray marr
+  interruptibleReceiveBoundedMutableByteArraySlice abandon conn sz marr off
```
Comment (andrewthad) [inline src/Socket/Stream/IPv4.hs]: This needs to be called in a loop until we receive exactly the requested number of bytes.

File: src/Socket/Stream/IPv4.hs (hunk 2 above)
Comment (chessai) [reply src/Socket/Stream/IPv4.hs]: isn't that the job of `interruptibleReceiveBoundedMutableByteArraySlice`?

File: src/Socket/Stream/IPv4.hs (hunk 3 above)
Comment (andrewthad) [reply src/Socket/Stream/IPv4.hs]: Also, `sizeofMutableByteArray` should be considered deprecated. Prefer `getSizeofMutableByteArray`.

File: src/Socket/Stream/IPv4.hs (hunk 2 above)
Comment (andrewthad) [reply src/Socket/Stream/IPv4.hs]: The `Bounded` variants receive up to a given number of bytes. The non-`Bounded` variants (which is what this function is) receive *exactly* a requested number of bytes by calling `recv` in a loop. Both variants take a length argument. There's a non-sliced variant that doesn't take a length argument, but that's not what is being defined here.

Comment (chessai) [reply src/Socket/Stream/IPv4.hs]: isn't the length of the slice just the (length<sub>bytearray</sub> - offset)?

Comment (chessai) [reply src/Socket/Stream/IPv4.hs]: oh, duh, slices don't always end at the end of the array. sorry.

**Comment by chessai** [comment]
changes resolved. requesting new review

File: src/Socket/Stream/IPv4.hs [hunk 4]
```diff
@@ -1280,6 +1283,62 @@ receiveBetweenMutableByteArraySlice !conn@(Connection fd) !minBytes !maxBytes !m
...
+--   completion of an 'STM' transaction.
+interruptibleReceiveMutableByteArraySlice ::
+     TVar Bool
+     -- ^ Interrupted. If this becomes 'True' give up and return
+     --   @'Left' 'ReceiveInterrupted'@.
+  -> Connection -- ^ Connection
+  -> MutableByteArray RealWorld -- ^ Buffer in which the data is going to be stored
+  -> Int -- ^ Offset into the buffer
+  -> Int -- ^ Length of slice
+  -> IO (Either (ReceiveException 'Interruptible) Int) -- ^ Either a socket exception or the number of bytes read
```
Comment (andrewthad) [inline src/Socket/Stream/IPv4.hs]: This should return `()`, not `Int`. If it succeeds at all, then `total_bytes === requested_bytes`

File: src/Socket/Stream/IPv4.hs [hunk 5]
```diff
@@ -1280,6 +1283,62 @@ receiveBetweenMutableByteArraySlice !conn@(Connection fd) !minBytes !maxBytes !m
...
+  let recvMax = interruptibleReceiveBoundedMutableByteArraySlice
+  let go off' remaining = case compare remaining 0 of
+        GT -> do
+          recvMax abandon conn remaining marr off' >>= \case
+            Left err -> pure (Left err)
+            Right sz' -> if sz' /= 0
+              then go (off' + sz') (remaining - sz')
+              else pure (Left ReceiveShutdown)
+        EQ -> do
+          pure (Right (sliceLen - off + 1))
```
Comment (andrewthad) [inline src/Socket/Stream/IPv4.hs]: This calculation is wrong (the correct value is just `sliceLen` without further arithmetic), but it's also unneeded for the reason mentioned above.

File: src/Socket/Stream/IPv4.hs [hunk 6]
```diff
@@ -1280,6 +1283,62 @@ receiveBetweenMutableByteArraySlice !conn@(Connection fd) !minBytes !maxBytes !m
...
       [SCK.negativeSliceLength]
 
+interruptibleReceiveBoundedByteStringSlice ::
+     TVar Bool
+     -- ^ Interrupted. If this becomes 'True' give up and return
+     --   @'Left' 'ReceiveInterrupted'@.
+  -> Connection -- ^ Connection
+  -> Int -- ^ Maximum number of bytes to receive
+  -> Int -- ^ Offset into the buffer
+  -> IO (Either (ReceiveException 'Interruptible) ByteString) -- ^ Either a socket exception or the number of ByteString we allocated
```
Comment (andrewthad) [inline src/Socket/Stream/IPv4.hs]: the haddock on this has a typo. It should read:

> Either a socket exception or the newly-allocated 'ByteString'

**Comment by chessai** [comment]
done. requesting another review

**Pull request merged**